File: 152-Maximum Product Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        }
        
        int maxHere = nums[0], minHere = nums[0], maxSofar = nums[0];
        for (int i = 1; i < nums.size(); ++i) {
            int mh = maxHere;
            maxHere = max(max(maxHere * nums[i], minHere * nums[i]), nums[i]);
            minHere = min(min(mh * nums[i], minHere * nums[i]), nums[i]);
            maxSofar = max(maxSofar, maxHere);
        }
        
        return maxSofar;
    }
};

File: 152-Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int result = nums[0];
        for (int i = 1, max = result, min = result; i < nums.length; ++i) {
            if (nums[i] < 0) {
                int temp = max;
                max = min;
                min = temp;
            }
            
            max = Math.max(nums[i], max * nums[i]);
            min = Math.min(nums[i], min * nums[i]);
            result = Math.max(result, max);
        }
        return result;
    }
}
